File: 241-differentWayToAddParentheses/solution.java
class Solution {
    public List<Integer> diffWaysToCompute(String input) {
        List<Integer> result= new ArrayList<>();
        
        // 50%
        for(int i=0;i<input.length(); i++){
            char c = input.charAt(i);
            if(!Character.isDigit(c)){
                List<Integer> left=diffWaysToCompute(input.substring(0, i));
                List<Integer> right=diffWaysToCompute(input.substring(i+1));
                
                for(int l: left){
                    for(int r: right){
                        switch(c){
                            case '+':
                                result.add(l+r);
                                break;
                            case '-':
                                result.add(l-r);
                                break;
                            case '*':
                                result.add(l*r);
                                break;
                            default:
                                result.add(l%r);
                        }
                    }
                }
            }
        }
        
        if(result.size()==0){
            result.add(Integer.valueOf(input));
        }
        
        return result;
    }
}

File: 241-differentWayToAddParentheses/solution.py
class Solution:
    def diffWaysToCompute(self, input):
        """
        :type input: str
        :rtype: List[int]
        """
        return self.traversal(self.parser(input))
        
    #    65%
    '''
        since using permutation have to deal with duplicate issue
        try another way of bottom-up approach
    '''
    def traversal(self, operation):
        # print(operation)
        result=[]
        
        if len(operation)==1:            
            return operation
        
        operator=set(['+', '-', '*', '/'])    
        
        for i in range(len(operation)):
            if operation[i] in operator:
                left=self.traversal(operation[:i])
                right=self.traversal(operation[i+1:])
                
                for operand1 in left:
                    for operand2 in right:
                        if operation[i]=='*':
                            result.append(operand1*operand2)
                        elif operation[i]=='+':
                            result.append(operand1+operand2)
                        elif operation[i]=='-':
                            result.append(operand1-operand2)
                        else:
                            result.append(operand1//operand2)                                                        
                
        return result
   
    def parser(self, chars):
        result=[]
        temp=""
        operator=set(['+', '-', '*', '/'])
        for c in chars:
            if not(c in operator):
                temp+=c
            else:
                result.append(temp)
                result.append(c)
                temp=""
                
        result.append(temp)
        if not(result[0].isnumeric()):
            result.insert(0, int(result[0]+result[1]))
            del result[1:3]
        return [int(x) if x.isnumeric() else x for x in result]
       